packages/sw-runtime-caching/test/sw/network-first.js

```diff
@@ -0,0 +1,80 @@
+importScripts('setup.js');
+
+describe('Test of the NetworkFirst handler', function() {
+  const CACHE_NAME = location.href;
+  const REQUEST_WRAPPER = new goog.runtimeCaching.RequestWrapper({cacheName: CACHE_NAME});
+  const NETWORK_FIRST = new goog.runtimeCaching.NetworkFirst({requestWrapper: REQUEST_WRAPPER, waitOnCache: true});
+  const COUNTER_URL = new URL('/__echo/counter', location).href;
+
+  let globalStubs = [];
+  let cache;
+  let initialCachedResponse;
+
+  before(() => {
+    const event = new FetchEvent('fetch', {request: new Request(COUNTER_URL)});
+    return caches.delete(CACHE_NAME)
+      .then(() => caches.open(CACHE_NAME))
+      .then((openedCache) => cache = openedCache)
+      .then(() => NETWORK_FIRST.handle({event}))
+      .then(() => cache.match(COUNTER_URL))
+      .then((cachedResponse) => initialCachedResponse = cachedResponse);
+  });
+
+  afterEach(function() {
+    globalStubs.forEach((stub) => stub.restore());
+    globalStubs = [];
+  });
+
+  it(`should add the initial response to the cache`, function() {
+    expect(initialCachedResponse).to.exist;
+  });
+
+  it(`should return the cached response and not update the cache when the network request fails`, function() {
+    globalStubs.push(sinon.stub(self, 'fetch').throws('NetworkError'));
+
+    const event = new FetchEvent('fetch', {request: new Request(COUNTER_URL)});
+    let handleResponse;
+    return NETWORK_FIRST.handle({event})
+      .then((response) => handleResponse = response)
+      .then(() => expectSameResponseBodies(initialCachedResponse, handleResponse))
+      .then(() => cache.match(COUNTER_URL))
+      .then((currentCachedResponse) => expectSameResponseBodies(currentCachedResponse, initialCachedResponse));
+  });
+
+  it(`should return the cached response if the network request times out`, function() {
```

Comment:
I haven't been able to successfully get this working with the asynchronous tests 😦 

I've lowered the amount of time that the code is sleeping for and it finishes quickly and doesn't appear to be flaky using the actual system clock, so I'm just going to leave as-is.